File: app/src/main/java/com/ed/redditapp/ui/subreddit/SubredditInfoAsyncTask.java
package com.ed.redditapp.ui.subreddit;

import com.ed.redditapp.lib.AsyncTask;
import com.ed.redditapp.lib.api.RedditApi;
import com.ed.redditapp.lib.api.SubReddit;

public class SubredditInfoAsyncTask {
    public interface Callback {
        void callback(SubReddit result);
    }

    private final RedditApi redditApi;

    public SubredditInfoAsyncTask(RedditApi redditApi) {
        this.redditApi = redditApi;
    }

    public void run(String subredditName, Callback callback) {
        new AsyncTask().runAsync((h) -> {
            SubReddit subreddit = redditApi.getSubredditInfo(subredditName);
            h.onUIThread(() -> callback.callback(subreddit));
        });
    }
}

File: app/src/main/java/com/ed/redditapp/ui/subreddit/SubredditInfoAsyncTask.kt
package com.ed.redditapp.ui.subreddit

import com.ed.redditapp.lib.AsyncTask
import com.ed.redditapp.lib.api.RedditApi
import com.ed.redditapp.lib.api.SubReddit

typealias SubredditInfoCallback = (result: SubReddit) -> Unit

class SubredditInfoAsyncTask(private val redditApi: RedditApi) {
    fun run(subredditName: String, callback: SubredditInfoCallback) {
        AsyncTask().runAsync { h ->
            val subreddit = redditApi.getSubredditInfo(subredditName)
            h.onUIThread { callback(subreddit) }
        }
    }
}